File: elo_model.py
# filename = elo_model.py
# author = John Brichetto
# purpose = Simulate a game between two teams (or individuals) comprised of players of random and unknown skill levels.
#           Accurately determine players relative skill level in as few games as possible.

import random

# Global Constants - Alter these variables to see how they affect an ELO estimation of a player's skill.

TOTAL_GAMES = 10000  # TOTAL number of games to be simulated - NOT the number of games each player will play.
TOTAL_PLAYERS = 1000  # Total number of players to simulate. TOTAL_PLAYERS MUST BE >= 2*TEAM_SIZE
TEAM_SIZE = 5  # The number of players on each team.
ELO_SPREAD = 400  # The Elo difference required for a player to be an order of magnitude more likely to win.
ELO_WEIGHT = 32  # Sets the "weight" of each match. Winning an even match causes a gain of 0.5 * ELO_WEIGHT
UN_RANKED_ELO = 1500.00  # The starting elo of all players.
DISPLAY_DATA_REPORT = False  # True will cause a printout of all players skill levels, elo score, and games played
DISPLAY_DATA_ANALYSIS = True  # True will cause a printout of the analysis


class Player:
    def __init__(self, name):
        self.name = name
        self.skill = random.randint(0, 10)
        self.elo = UN_RANKED_ELO
        self.games_played = 0

    def elo_change(self, change):
        self.elo += change
        self.games_played += 1

    def get_skill(self): return self.skill

    def get_elo(self): return self.elo

    def get_name(self): return self.name

    def get_games_played(self): return self.games_played


# Create Players
player_pool = []
for i in range(0, TOTAL_PLAYERS):
    player_pool.append(Player("p"+str(i)))

# Create two empty teams
team1 = []
team2 = []

for j in range(0, TOTAL_GAMES):
    # Fill the teams with players
    while len(team1) < TEAM_SIZE:
        team1.append(player_pool.pop(random.randint(0, (len(player_pool) - 1))))
        team2.append(player_pool.pop(random.randint(0, (len(player_pool) - 1))))

    # Initialize / Reset elo and skill
    team1_skill, team1_elo = 0, 0
    team2_skill, team2_elo = 0, 0

    # Calculate each team's skill and average elo
    for i in range(0, len(team1)):
        team1_elo += team1[i].get_elo()
        team1_skill += team1[i].get_skill()
        team2_elo += team2[i].get_elo()
        team2_skill += team2[i].get_skill()
    team1_elo = team1_elo / len(team1)
    team2_elo = team2_elo / len(team2)

    # Determine tug of war winner
    if team1_skill > team2_skill:
        team1_win = 1.0
    elif team1_skill == team2_skill:
        team1_win = 0.5
    else:
        team1_win = 0.0

    # Calculate elo change
    n = ELO_SPREAD  # n defines the elo difference needed to be an order of magnitude more likely to win
    k = ELO_WEIGHT  # k defines the weight of each match
    r1 = 10**(team1_elo / n)
    r2 = 10**(team2_elo / n)
    t1_win_prob = r1 / (r1+r2)
    t2_win_prob = r2 / (r1+r2)
    t1_elo_change = k*(team1_win - t1_win_prob)  # if team 1 wins, this creates a positive number (elo gain)
    t2_elo_change = k*(abs(team1_win - 1.0) - t2_win_prob)  # if team 1 wins, this creates a negative number (elo loss)

    # Apply elo change to each player and update games played
    for i in range(0, len(team1)):
        team1[i].elo_change(t1_elo_change)
        team2[i].elo_change(t2_elo_change)

    # Return Players to the player_pool
    for i in range(0, len(team1)):
        player_pool.append(team1.pop())
        player_pool.append(team2.pop())

# Organize the Player Pool
if DISPLAY_DATA_REPORT:
    organization_counter = 0
    for i in range(0, 10):
        for j in range(0, len(player_pool) - 1):
            if player_pool[j].get_skill() == (10 - i):
                player_pool.insert(organization_counter, player_pool.pop(j))
                organization_counter += 1
            else:
                pass

# Report Data
if DISPLAY_DATA_REPORT:
    for i in range(0, len(player_pool) - 1):
        print("Name: ", player_pool[i].get_name(), "\t\tSkill: ", player_pool[i].get_skill(),
              "\t\tElo : %.2f" % player_pool[i].get_elo(), "\t\tGames Played:  ", player_pool[i].get_games_played())
else:
    pass

# Analyze Data
if DISPLAY_DATA_ANALYSIS:
    for i in range(0, 11):
        analysis_elo_list = []
        analysis_games_played_list = []
        for j in range(0, len(player_pool) - 1):
            if player_pool[j].get_skill() == (10 - i):
                analysis_elo_list.append(int(player_pool[j].get_elo()))
                analysis_games_played_list.append(player_pool[j].get_games_played())
                analysis_elo_list.sort()
            else:
                pass

        print("Skill Level: ", (10 - i),
              "\tHighest Elo: ", analysis_elo_list[(len(analysis_elo_list)-1)],
              "\tLowest Elo: ", analysis_elo_list[0],
              "\tAverage Elo: %.2f" % (sum(analysis_elo_list) / len(analysis_elo_list)),
              "\tPlayers this skill level: ", len(analysis_elo_list),
              "\tAverage Number of Games Played: %.2f"
              % (sum(analysis_games_played_list) / len(analysis_games_played_list))
              )

File: elo_model.go
package main

/*
PURPOSE:
This program was created to demonstrate the Elo sorting method's ability
to estimate an individual player's skill level based on their performance
in a team game.

EXAMPLE:
If two teams play tug-of-war against each other, is it possible to tell
if every player on the winning team was just slightly stronger? or maybe
only one person was significantly stronger?

AUTHOR: John Brichetto
*/

import (
	"fmt"
	"math"
	"math/rand"
	"strconv"
	"time"
)

// game related constant declaration
const totalGamesPlayed int = 10000
const totalPlayers int = 1000
const teamSize int = 5

// player related constant declaration
const maxSkillLevel int = 10
const initialElo float64 = 1500.00

// Elo algorithm related constant declaration
const eloSpread float64 = 400.00
const eloWeight float64 = 32.00

// output related constants
const displayAllPlayers bool = false
const displaySummary bool = true

type player struct {
	name        string
	skill       int
	elo         float64
	gamesPlayed int
}

func playerConstructor(inputName string) player {
	p := player{
		name:        inputName,
		skill:       rand.Intn(maxSkillLevel + 1),
		elo:         initialElo,
		gamesPlayed: 0,
	}
	return p
}

func playerRemove(p []player, i int) []player {
	p[i] = p[len(p)-1]
	return p[:len(p)-1]
}

func stdDev(nums ...float64) (float64, float64) {
	sum := 0.00
	for _, num := range nums {
		sum += num
	}
	average := sum / float64(len(nums))
	var sD float64
	for _, num := range nums {
		sD += math.Pow((num - average), 2)
	}
	sD = math.Sqrt(sD / float64(len(nums)))
	return average, sD
}

func main() {
	// begin timing
	startTime := time.Now()
	defer func() {
		elapsedTime := time.Since(startTime)
		fmt.Println(elapsedTime)
	}()

	//create player pool
	playerPool := make([]player, totalPlayers, totalPlayers+1)
	for i := 0; i < totalPlayers; i++ {
		playerPool[i] = playerConstructor("p" + strconv.Itoa(i))
	}

	// simulate games
	for gameCounter := 0; gameCounter < totalGamesPlayed; gameCounter++ {

		// create and fill teams
		team1 := make([]player, 0, teamSize)
		for i := 0; i < teamSize; i++ {
			randomPlayerID := rand.Intn(len(playerPool))
			team1 = append(team1, playerPool[randomPlayerID])
			playerPool = playerRemove(playerPool, randomPlayerID)
		}
		team2 := make([]player, 0, teamSize)
		for i := 0; i < teamSize; i++ {
			randomPlayerID := rand.Intn(len(playerPool))
			team2 = append(team2, playerPool[randomPlayerID])
			playerPool = playerRemove(playerPool, randomPlayerID)
		}
		// calculate the team's average Elo and determine which team is more skilled
		team1AvgElo, team2AvgElo := 0.00, 0.00
		team1Skill, team2Skill := 0, 0
		for i := 0; i < teamSize; i++ {
			team1AvgElo = team1AvgElo + team1[i].elo
			team1Skill += team1[i].skill
			team2AvgElo = team2AvgElo + team2[i].elo
			team2Skill += team2[i].skill
		}
		team1AvgElo /= float64(teamSize)
		team2AvgElo /= float64(teamSize)

		// determine game winner by team's total skill
		var team1Win float64
		if team1Skill > team2Skill {
			team1Win = 1.00
		} else if team1Skill == team2Skill {
			team1Win = 0.50
		} else {
			team1Win = 0.00
		}

		// calculations
		r1 := math.Pow(10, team1AvgElo/eloSpread)
		r2 := math.Pow(10, team2AvgElo/eloSpread)

		team1WinProbability := r1 / (r1 + r2)
		team2WinProbability := r2 / (r1 + r2)

		// if team 1 wins, this creates a positive number (elo gain)
		team1EloChange := eloWeight * (team1Win - team1WinProbability)
		// if team 1 wins, this creates a negative number (elo loss)
		team2EloChange := eloWeight * (math.Abs(team1Win-1.00) - team2WinProbability)

		// update elo score and games played for each player and return to playerPool
		for i := (teamSize - 1); i >= 0; i-- {
			team1[i].elo += team1EloChange
			team1[i].gamesPlayed++
			playerPool = append(playerPool, team1[i])
			team1 = playerRemove(team1, i)
			team2[i].elo += team2EloChange
			team2[i].gamesPlayed++
			playerPool = append(playerPool, team2[i])
			team2 = playerRemove(team2, i)
		}
	}

	// organize the player pool
	sortedPlayerPool := make([]player, 0, totalPlayers)
	for j := 0; j <= maxSkillLevel; j++ {
		for k := 0; k < len(playerPool); k++ {
			if playerPool[k].skill == (maxSkillLevel - j) {
				sortedPlayerPool = append(sortedPlayerPool, playerPool[k])
				playerPool = playerRemove(playerPool, k)
				k--
			}
		}
	}

	//display the sorted player pool
	if displayAllPlayers {
		for i := 0; i < len(sortedPlayerPool); i++ {
			fmt.Println(sortedPlayerPool[i])
		}
	}

	// summarize the data - calculate relevant statistical data for each skill level
	if displaySummary {
		for j := 0; j <= maxSkillLevel; j++ {
			eloScoresThisSkillLevel := make([]float64, 0, totalPlayers)
			maxVal := 0.00
			minVal := 1000000.00
			for len(sortedPlayerPool) != 0 {
				if sortedPlayerPool[0].skill == (maxSkillLevel - j) {
					eloScoresThisSkillLevel = append(eloScoresThisSkillLevel, sortedPlayerPool[0].elo)
					maxVal = math.Max(maxVal, sortedPlayerPool[0].elo)
					minVal = math.Min(minVal, sortedPlayerPool[0].elo)
					sortedPlayerPool = sortedPlayerPool[1:]
				} else {
					avg, sD := stdDev(eloScoresThisSkillLevel...)
					fmt.Println("Skill Level:", maxSkillLevel-j, "\tNumber of Players:", len(eloScoresThisSkillLevel), "\tAverage elo:", avg, "\tStandard Deviation:", sD,
						"\tMax elo:", maxVal, "\tMinimum elo:", minVal)
					break
				}
			}
			if len(sortedPlayerPool) == 0 {
				avg, sD := stdDev(eloScoresThisSkillLevel...)
				fmt.Println("Skill Level:", maxSkillLevel-j, "\tNumber of Players:", len(eloScoresThisSkillLevel), "\tAverage elo:", avg, "\tStandard Deviation:", sD,
					"\tMax elo:", maxVal, "\tMinimum elo:", minVal)
			}
		}
	}
}
